### PR TITLE
10939 make affected tables private

### LINKED ISSUE #10939: Make Layer.affected_tables private
- `Layer.affected_tables` calculates the affected tables from the query, analyses, etc.
- `Layer.register_table_dependencies` takes the previous output and writes it to the DB (layers_user_table)
- `Layer.user_table` returns the cached version of the affected tables from the DB

I think we should make `affected_tables` private to enforce used of the cached version. Right now, `affected_tables` is being used in handful of places around the code (e.g: vizjson3 presenter, .carto exporter) but it makes no sense to use this instead of the caching version (assuming the cached version is correct, which we should strive to be the case).

cc @CartoDB/builder-backend opinions?

### PR DESCRIPTION
**Pull request by javitonino**
Closes #10939

Starts using the cached version of layer<->table dependencies for everything. Since this depends on the cache, add tests to ensure we regenerate it as needed:
- [x] Layer updates
- [x] Analyses updates
- [x] Importer (including .carto)
- [x] Duplicate map

**Acceptance** needs to looks for issues where the table dependencies cache is not correctly updated. Acceptance would have failed if you can find some way in which the user tables are not updated. You can check that from the UI (trying to delete a dataset should warn about the visualization being affected), or from the rails console:
`Carto::Visualization.find(<viz_id>).map.layers.flat_map(&:user_tables).uniq` should be accurate (be careful reloading).

In order to cause this issues, try creating/updating/deleting a map in any way you can think of (import, from empty layer, duplicate, from sql query), all kinds of data layers (empty, from existing tables, from import). Also, test analyses in depth, as those can add dependencies as well.

**Deploy**
Aside from deploying, we must also run the update task. `cartodb:db:register_table_dependencies`

### CODE REVIEW

**Comment by juanignaciosl** [comment]
Before the CR itself, you did [two interesting comments in the issue](https://github.com/CartoDB/cartodb/issues/10939#issuecomment-266404420):

- "I would consider running a rake to rebuild all caches.": how come? If the cache is going to be filled as the tables are requested, there's no hurry to warm it up.
- "I would like to do some benchmarking of cached vs uncached first": did you?

File: app/models/carto/layer.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 module Carto
   module LayerTableDependencies
+    private
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: This is probably wrong

**Comment by javitonino** [comment]
Answering:
 - We have to run the rake, because the current code rebuilds the caches on any update, but not on any request (it wouldn't make sense to rebuild on GETs). So if we don't rebuild it, caches that are not correct now would be a problem if a person tries to delete a dataset before touching all the maps it's being used in.
 - Benchmark. I tested the time it took to run `user_tables` (query to cache) vs `affected_tables` (calculation) in all maps under the `team` org. The user table was 21s. `affected_tables` took 5min (310s). This is for 10k maps, so for a single map, the impact of `affected_tables` is 30ms, and for `user_tables` 2 ms. An order of magnitude.

File: app/models/carto/layer.rb (hunk 1 above)
Comment (javitonino) [inline app/models/carto/layer.rb]: That's the flag change of this PR, it makes `affected_tables` private. I checked that it works as intended inside an included module.

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: But in that case the next `private` in this file should be removed, isn't it?

Comment (javitonino) [inline app/models/carto/layer.rb]: This is inside a mixin (`Carto::LayerTableDependencies`) which is defined in the same file as the model (`Carto::Layer`). Each of those has its separate `private`.

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: My bad! 1€

**Pull request merged**